libs/core/langchain_core/runnables/graph.py

```diff
@@ -25,22 +35,108 @@ class Node(NamedTuple):
     data: Union[Type[BaseModel], RunnableType]
 
 
+def node_data_str(node: Node) -> str:
+    from langchain_core.runnables.base import Runnable
+
+    if not is_uuid(node.id):
+        return node.id
+    elif isinstance(node.data, Runnable):
+        try:
+            data = str(node.data)
+            if (
+                data.startswith("<")
+                or data[0] != data[0].upper()
+                or len(data.splitlines()) > 1
+            ):
+                data = node.data.__class__.__name__
+            elif len(data) > 42:
+                data = data[:42] + "..."
+        except Exception:
+            data = node.data.__class__.__name__
+    else:
+        data = node.data.__name__
+    return data if not data.startswith("Runnable") else data[8:]
+
+
+def node_data_json(node: Node) -> Dict[str, Union[str, Dict[str, Any]]]:
+    from langchain_core.load.serializable import to_json_not_implemented
+    from langchain_core.runnables.base import Runnable, RunnableSerializable
+
+    if isinstance(node.data, RunnableSerializable):
+        return {
+            "type": "runnable",
+            "data": {
+                "id": node.data.lc_id(),
+                "name": node.data.get_name(),
+            },
+        }
+    elif isinstance(node.data, Runnable):
+        return {
+            "type": "runnable",
+            "data": {
+                "id": to_json_not_implemented(node.data)["id"],
+                "name": node.data.get_name(),
+            },
+        }
+    elif inspect.isclass(node.data) and issubclass(node.data, BaseModel):
+        return {
+            "type": "schema",
+            "data": node.data.schema(),
+        }
+    else:
+        return {
+            "type": "unknown",
+            "data": node_data_str(node),
+        }
+
+
 @dataclass
 class Graph:
     """Graph of nodes and edges."""
 
     nodes: Dict[str, Node] = field(default_factory=dict)
     edges: List[Edge] = field(default_factory=list)
 
+    def to_json(self) -> Dict[str, List[Dict[str, Any]]]:
+        """Convert the graph to a JSON-serializable format."""
+        stable_node_ids = {
```

Comment:
what does stable mean here

Reply:
ah ic, so that two graphs with equivalent objects and edges would have same serialization?

Reply:
basically if you generate the same chain twice you should get same json graph